### ISSUE TITLE
multi-cpu problem in experiment grid

### ISSUE THREAD

**Issue by watchernyu** [openai/spinningup#74]
@jachiam Hi! It's me again! 2 days ago I posted an issue on using multiple cpu on ExperimentGrid that seems to only give the wrong log when run in Pycharm but fine in terminal. I did some more experiments today, and found that although when run from terminal, the log seems to be correct, but there could still be a bug, the proof is that a multi-cpu run will always use more time compared to single cpu run, when solving the same number of jobs.

Here is the code I tried, when you have time, could you try run it and see if you can reproduce the bug?
My machine has 4 cores, and using 2 cores to run 4 jobs should be faster than using 1 core to run 4 jobs. 
The result I get is: single-cpu about 15 seconds, multi-cpu about 20 seconds.
I have run this code from terminal, the result indicates that with the same number of jobs, multi-cpu uses more time than single-cpu, which seems a bug to me.
I believe what the code might be doing is that for every variant, the same variant has been run num_cpu times.

```
from spinup.utils.run_utils import ExperimentGrid
from spinup import ppo

import time
if __name__ == '__main__':
    total_num_runs = 4
    import argparse

    ## try multi-cpu case
    parser = argparse.ArgumentParser()
    parser.add_argument('--cpu', type=int, default=2)
    parser.add_argument('--num_runs', type=int, default=total_num_runs)
    args = parser.parse_args()

    ## reset timing
    starttime = time.time()
    eg = ExperimentGrid(name='ppo-bench')
    eg.add('env_name', 'CartPole-v0', '', True)
    eg.add('seed', [10 * i for i in range(args.num_runs)])
    eg.add('epochs', 1)
    eg.add('steps_per_epoch', 200)
    eg.run(ppo, num_cpu=args.cpu)
    multi_cpu_time = time.time() - starttime

    ## try single-cpu case
    args.cpu = 1

    ## reset timing
    starttime = time.time()
    eg = ExperimentGrid(name='ppo-bench')
    eg.add('env_name', 'CartPole-v0', '', True)
    eg.add('seed', [10 * i for i in range(args.num_runs)])
    eg.add('epochs', 1)
    eg.add('steps_per_epoch', 200)
    eg.run(ppo, num_cpu=args.cpu)
    single_cpu_time = time.time() - starttime

    print('single-cpu',single_cpu_time,'multi-cpu',multi_cpu_time)
```
Many thanks!

**Comment by jachiam** [comment]
Hi @watchernyu! Not sure if these results are too terribly surprising, or if they indicate a bug. There's a trade-off in compute time for parallelized programs, because you have to incur some additional time cost for communicating information between processes. If the communications overhead is greater than the amount of time you would gain from the part you parallelize, you will get a slowdown instead of a speedup. In your tests, you are using a very small batch size (`steps_per_eoch`), which probably puts you in the regime where parallelization hurts more than it helps. Check with a much larger batch size and see if the issue still holds.

**Comment by jachiam** [comment]
That is expected behavior! So, `num_cpu` doesn't parallelize the jobs: it parallelizes each single run of PPO.

Jobs are still run sequentially. In each job, the launcher spawns `num_cpu` processes, each of which simultaneously runs the `ppo` function. Inside the `ppo` function, a call to `sync_all_params` makes sure that all the processes talk to each other and have the same randomly-initialized neural network parameters to start with. Then, each gradient descent step with the `MpiAdamOptimizer` averages gradients across the processes, to ensure that they all update the same way. This keeps the agent synced across processes throughout training.

The parallelization comes in handy because it speeds up environment interactions: instead of a single process running a full `steps_per_epoch` number of steps in sequence, each process runs `steps_per_epoch / num_cpu` steps in sequence.

But if `steps_per_epoch` is small to begin with, and a single environment step is fast, this doesn't tend to speed you up very much.

**Comment by watchernyu** [comment]
Thanks for the explanation! Maybe I got it all wrong... is this also the expected behavior for other algorithms such as SAC?

**Comment by sandipan1** [comment]
Is there a reason why no parallelization is supported for DDPG, TD3, SAC?  Or what is the simplest way to support parallelization for these off-policy algorithms?
